# Import thÆ° viá»‡n gRPC
import grpc
# Import concurrent.futures Ä‘á»ƒ xá»­ lÃ½ multi-threading
from concurrent import futures
# Import sys Ä‘á»ƒ láº¥y command line arguments (port number)
import sys
# Import os Ä‘á»ƒ xá»­ lÃ½ paths
import os

# ThÃªm project root vÃ o Python path Ä‘á»ƒ import Ä‘Æ°á»£c module local
project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, project_root)

# Import generated code tá»« protobuf
# kvstore_pb2: Chá»©a message classes (PutRequest, GetResponse, etc.)
from src.proto import kvstore_pb2
# kvstore_pb2_grpc: Chá»©a service classes (Servicer, Stub)
from src.proto import kvstore_pb2_grpc


# Class nÃ y káº¿ thá»«a tá»« KeyValueStoreServicer (generated tá»« proto)
# Servicer lÃ  base class cho server implementation
class KeyValueStoreServicer(kvstore_pb2_grpc.KeyValueStoreServicer):
    """
    Implementation cá»§a KeyValueStore service
    Xá»­ lÃ½ cÃ¡c requests tá»« clients: PUT, GET, DELETE, ListKeys
    """
    
    def __init__(self):
        """Constructor - khá»Ÿi táº¡o server"""
        # TODO: Phase 2 sáº½ thÃªm storage engine vÃ o Ä‘Ã¢y
        print("âœ… KeyValueStoreServicer initialized")
    
    # Override method Put tá»« base class
    # request: PutRequest object (cÃ³ key, value fields)
    # context: gRPC context (metadata, authentication, etc.)
    def Put(self, request, context):
        """Handler cho PUT operation"""
        # TODO: Phase 2 sáº½ implement logic lÆ°u data
        print(f"ðŸ“¥ Received PUT request: key={request.key}")
        
        # Tráº£ vá» empty response (chá»‰ Ä‘á»ƒ test Phase 1)
        return kvstore_pb2.PutResponse()
    
    # Override method Get
    def Get(self, request, context):
        """Handler cho GET operation"""
        # TODO: Phase 2 sáº½ implement logic Ä‘á»c data
        print(f"ðŸ“¤ Received GET request: key={request.key}")
        
        # Tráº£ vá» empty response
        return kvstore_pb2.GetResponse()
    
    # Override method Delete
    def Delete(self, request, context):
        """Handler cho DELETE operation"""
        # TODO: Phase 2 sáº½ implement logic xÃ³a data
        print(f"ðŸ—‘ï¸ Received DELETE request: key={request.key}")
        
        # Tráº£ vá» empty response
        return kvstore_pb2.DeleteResponse()
    
    # Override method ListKeys
    def ListKeys(self, request, context):
        """Handler cho ListKeys operation - list táº¥t cáº£ keys"""
        # TODO: Phase 2 sáº½ implement logic list keys
        print("ðŸ“‹ Received ListKeys request")
        
        # Tráº£ vá» empty response
        return kvstore_pb2.ListKeysResponse()


def serve(port):
    """
    HÃ m chÃ­nh Ä‘á»ƒ start gRPC server
    
    Args:
        port: Port number Ä‘á»ƒ server listen (vd: 8001)
    """
    # Táº¡o gRPC server vá»›i thread pool
    # max_workers=10: Tá»‘i Ä‘a 10 threads xá»­ lÃ½ requests Ä‘á»“ng thá»i
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    
    # ÄÄƒng kÃ½ servicer vÃ o server
    # add_KeyValueStoreServicer_to_server: Generated function tá»« proto
    kvstore_pb2_grpc.add_KeyValueStoreServicer_to_server(
        KeyValueStoreServicer(),  # Instance cá»§a servicer class
        server  # Server object
    )
    
    # Bind server vÃ o address
    # [::]:port nghÄ©a lÃ  listen trÃªn táº¥t cáº£ network interfaces
    server_address = f'[::]:{port}'
    server.add_insecure_port(server_address)
    
    # Start server
    server.start()
    print(f"ðŸš€ Server started on port {port}")
    print(f"ðŸ“¡ Listening on {server_address}")
    print("Press Ctrl+C to stop")
    
    # Giá»¯ server cháº¡y cho Ä‘áº¿n khi Ctrl+C
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("\nâ¹ï¸ Server stopping...")
        server.stop(0)  # Grace period 0 giÃ¢y
        print("âœ… Server stopped")


# Entry point cá»§a script
if __name__ == '__main__':
    # Láº¥y port tá»« command line arguments
    # Default port = 8001 náº¿u khÃ´ng truyá»n vÃ o
    if len(sys.argv) > 1:
        port = int(sys.argv[1])  # Parse string -> int
    else:
        port = 8001  # Default port
    
    print(f"ðŸŽ¯ Starting server on port {port}...")
    serve(port)